File: packages/fleet-back/src/routers/analyst/analytics.ts
import { and, eq, gte, sql, sum } from "drizzle-orm";
import * as v from "valibot";
import {
	journalEntryCreditItems,
	journalEntryDebitItems,
} from "@/db/analytics-schema";
import { expenseCategory, vehicles } from "@/db/schema";
import { analystAuthorized } from "@/procedures/analyst-authorized";
import { vehicleAnalyticsRouter } from "@/routers/analyst/analytics/vehicle";

const periodSchema = v.picklist([
	"all_time",
	"last_7d",
	"last_30d",
	"last_6m",
	"last_12m",
]);

type Period = v.InferOutput<typeof periodSchema>;

function getPeriodDates(period: Period): {
	currentStart: Date | null;
	previousStart: Date | null;
	previousEnd: Date | null;
} {
	if (period === "all_time") {
		return { currentStart: null, previousStart: null, previousEnd: null };
	}

	const now = new Date();
	const currentStart = new Date();
	const previousStart = new Date();
	const previousEnd = new Date();

	switch (period) {
		case "last_7d":
			currentStart.setDate(now.getDate() - 7);
			previousEnd.setDate(now.getDate() - 7);
			previousStart.setDate(now.getDate() - 14);
			break;
		case "last_30d":
			currentStart.setDate(now.getDate() - 30);
			previousEnd.setDate(now.getDate() - 30);
			previousStart.setDate(now.getDate() - 60);
			break;
		case "last_6m":
			currentStart.setMonth(now.getMonth() - 6);
			previousEnd.setMonth(now.getMonth() - 6);
			previousStart.setMonth(now.getMonth() - 12);
			break;
		case "last_12m":
			currentStart.setMonth(now.getMonth() - 12);
			previousEnd.setMonth(now.getMonth() - 12);
			previousStart.setMonth(now.getMonth() - 24);
			break;
	}

	return { currentStart, previousStart, previousEnd };
}

function calculatePercentageChange(
	current: number,
	previous: number,
): number | null {
	if (previous === 0) {
		return current > 0 ? 100 : null;
	}
	return ((current - previous) / previous) * 100;
}

const summaryStats = analystAuthorized
	.input(v.object({ period: periodSchema }))
	.handler(async ({ input, context: { db } }) => {
		const { currentStart, previousStart, previousEnd } = getPeriodDates(
			input.period,
		);

		// Build conditions for current period
		const currentCreditConditions =
			currentStart !== null
				? and(gte(journalEntryCreditItems.transactionDate, currentStart))
				: undefined;

		const currentDebitConditions =
			currentStart !== null
				? and(gte(journalEntryDebitItems.transactionDate, currentStart))
				: undefined;

		// Build conditions for previous period
		const previousCreditConditions =
			previousStart !== null && previousEnd !== null
				? and(
						gte(journalEntryCreditItems.transactionDate, previousStart),
						sql`${journalEntryCreditItems.transactionDate} < ${previousEnd}`,
					)
				: undefined;

		const previousDebitConditions =
			previousStart !== null && previousEnd !== null
				? and(
						gte(journalEntryDebitItems.transactionDate, previousStart),
						sql`${journalEntryDebitItems.transactionDate} < ${previousEnd}`,
					)
				: undefined;

		// Execute all queries in parallel
		const [
			currentCreditResult,
			currentDebitResult,
			previousCreditResult,
			previousDebitResult,
		] = await Promise.all([
			// Current period revenue (credits)
			db
				.select({
					total: sum(journalEntryCreditItems.amount),
				})
				.from(journalEntryCreditItems)
				.where(currentCreditConditions),

			// Current period expenses (debits)
			db
				.select({
					total: sum(journalEntryDebitItems.amount),
				})
				.from(journalEntryDebitItems)
				.where(currentDebitConditions),

			// Previous period revenue (credits) - only if not all_time
			input.period !== "all_time"
				? db
						.select({
							total: sum(journalEntryCreditItems.amount),
						})
						.from(journalEntryCreditItems)
						.where(previousCreditConditions)
				: Promise.resolve([{ total: null }]),

			// Previous period expenses (debits) - only if not all_time
			input.period !== "all_time"
				? db
						.select({
							total: sum(journalEntryDebitItems.amount),
						})
						.from(journalEntryDebitItems)
						.where(previousDebitConditions)
				: Promise.resolve([{ total: null }]),
		]);

		// Parse results
		const currentRevenue = Number(currentCreditResult[0]?.total ?? 0);
		const currentExpenses = Number(currentDebitResult[0]?.total ?? 0);
		const currentProfit = currentRevenue - currentExpenses;
		const currentProfitPercentage =
			currentRevenue > 0 ? (currentProfit / currentRevenue) * 100 : 0;

		const previousRevenue = Number(previousCreditResult[0]?.total ?? 0);
		const previousExpenses = Number(previousDebitResult[0]?.total ?? 0);
		const previousProfit = previousRevenue - previousExpenses;
		const previousProfitPercentage =
			previousRevenue > 0 ? (previousProfit / previousRevenue) * 100 : 0;

		// Calculate percentage changes
		const revenueChange = calculatePercentageChange(
			currentRevenue,
			previousRevenue,
		);
		const expensesChange = calculatePercentageChange(
			currentExpenses,
			previousExpenses,
		);
		const profitChange = calculatePercentageChange(
			currentProfit,
			previousProfit,
		);
		const profitPercentageChange = calculatePercentageChange(
			currentProfitPercentage,
			previousProfitPercentage,
		);

		return {
			revenue: {
				value: currentRevenue,
				change: revenueChange,
			},
			expenses: {
				value: currentExpenses,
				change: expensesChange,
			},
			profit: {
				value: currentProfit,
				change: profitChange,
			},
			profitPercentage: {
				value: currentProfitPercentage,
				change: profitPercentageChange,
			},
		};
	});

const fleetStats = analystAuthorized
	.input(v.object({ period: periodSchema }))
	.handler(async ({ input, context: { db } }) => {
		const { currentStart } = getPeriodDates(input.period);

		// Build date condition
		const creditDateCondition =
			currentStart !== null
				? gte(journalEntryCreditItems.transactionDate, currentStart)
				: undefined;

		const debitDateCondition =
			currentStart !== null
				? gte(journalEntryDebitItems.transactionDate, currentStart)
				: undefined;

		// Get credit totals per vehicle
		const creditQuery = db
			.select({
				vehicleId: journalEntryCreditItems.vehicleId,
				vehicleName: vehicles.name,
				total: sum(journalEntryCreditItems.amount),
			})
			.from(journalEntryCreditItems)
			.innerJoin(vehicles, eq(journalEntryCreditItems.vehicleId, vehicles.id))
			.where(creditDateCondition)
			.groupBy(journalEntryCreditItems.vehicleId, vehicles.name);

		// Get debit totals per vehicle
		const debitQuery = db
			.select({
				vehicleId: journalEntryDebitItems.vehicleId,
				vehicleName: vehicles.name,
				total: sum(journalEntryDebitItems.amount),
			})
			.from(journalEntryDebitItems)
			.innerJoin(vehicles, eq(journalEntryDebitItems.vehicleId, vehicles.id))
			.where(debitDateCondition)
			.groupBy(journalEntryDebitItems.vehicleId, vehicles.name);

		// Get all vehicles (to include those with no transactions)
		const allVehiclesQuery = db
			.select({
				id: vehicles.id,
				name: vehicles.name,
				color: vehicles.color,
			})
			.from(vehicles);

		const [creditResults, debitResults, allVehicles] = await Promise.all([
			creditQuery,
			debitQuery,
			allVehiclesQuery,
		]);

		// Create maps for easy lookup
		const creditMap = new Map(
			creditResults.map((r) => [r.vehicleId, Number(r.total ?? 0)]),
		);
		const debitMap = new Map(
			debitResults.map((r) => [r.vehicleId, Number(r.total ?? 0)]),
		);

		// Combine results for all vehicles
		const vehicleData = allVehicles.map((vehicle) => {
			const credit = creditMap.get(vehicle.id) ?? 0;
			const debit = debitMap.get(vehicle.id) ?? 0;
			const profit = credit - debit;

			return {
				vehicleId: vehicle.id,
				vehicleName: vehicle.name,
				vehicleColor: vehicle.color,
				credit,
				debit,
				profit,
			};
		});

		// Sort by profit descending (best performing first)
		vehicleData.sort((a, b) => b.profit - a.profit);

		return vehicleData;
	});

const expensesStats = analystAuthorized
	.input(v.object({ period: periodSchema }))
	.handler(async ({ input, context: { db } }) => {
		const { currentStart } = getPeriodDates(input.period);

		// Build date condition
		const dateCondition =
			currentStart !== null
				? gte(journalEntryDebitItems.transactionDate, currentStart)
				: undefined;

		// Get expense totals per category
		const results = await db
			.select({
				categoryId: expenseCategory.id,
				categoryName: expenseCategory.name,
				categoryColor: expenseCategory.color,
				total: sum(journalEntryDebitItems.amount),
			})
			.from(journalEntryDebitItems)
			.innerJoin(
				expenseCategory,
				eq(journalEntryDebitItems.expenseCategoryId, expenseCategory.id),
			)
			.where(dateCondition)
			.groupBy(expenseCategory.id, expenseCategory.name, expenseCategory.color);

		// Transform results
		const categoryData = results.map((row) => ({
			id: row.categoryId,
			name: row.categoryName,
			color: row.categoryColor,
			amount: Number(row.total ?? 0),
		}));

		// Sort by amount descending (highest expenses first)
		categoryData.sort((a, b) => b.amount - a.amount);

		return categoryData;
	});

export const analyticsRouter = {
	summaryStats,
	fleetStats,
	expensesStats,
	vehicle: vehicleAnalyticsRouter,
};

File: packages/fleet-back/fleet_backend/routes/analytics.py
from typing import Any

from flask import jsonify, request

from database import connect, rows_to_dicts
from fleet_backend.common import (
    PERIODS,
    VEHICLE_PERIODS,
    calculate_percentage_change,
    period_bounds,
    period_start,
    require_auth,
    rpc_payload,
    rpc_response,
    sum_amount,
    vehicle_period_bounds,
    vehicle_period_start,
)
from fleet_backend.server import app

@app.get("/api/analytics/summary")
@require_auth({"analyst"})
def analytics_summary(user):
    period = request.args.get("period", "all_time")
    if period not in PERIODS:
        return jsonify({"error": "Invalid period"}), 400
    current_start, previous_start, previous_end = period_bounds(period)
    with connect() as conn:
        current_revenue = sum_amount(conn, "credit", start=current_start)
        current_expenses = sum_amount(conn, "debit", start=current_start)
        previous_revenue = sum_amount(conn, "credit", start=previous_start, end=previous_end)
        previous_expenses = sum_amount(conn, "debit", start=previous_start, end=previous_end)

    current_profit = current_revenue - current_expenses
    current_profit_percentage = (current_profit / current_revenue * 100) if current_revenue else 0
    previous_profit = previous_revenue - previous_expenses
    previous_profit_percentage = (previous_profit / previous_revenue * 100) if previous_revenue else 0

    return jsonify({
        "revenue": {
            "value": current_revenue,
            "change": calculate_percentage_change(current_revenue, previous_revenue),
        },
        "expenses": {
            "value": current_expenses,
            "change": calculate_percentage_change(current_expenses, previous_expenses),
        },
        "profit": {
            "value": current_profit,
            "change": calculate_percentage_change(current_profit, previous_profit),
        },
        "profitPercentage": {
            "value": current_profit_percentage,
            "change": calculate_percentage_change(current_profit_percentage, previous_profit_percentage),
        },
    })


@app.post("/orpc/analyst/analytics/summaryStats")
@app.post("/api/orpc/analyst/analytics/summaryStats")
@app.post("/analyst/analytics/summaryStats")
@require_auth({"analyst"})
def orpc_analytics_summary_stats(user):
    period = rpc_payload().get("period", "all_time")
    if period not in PERIODS:
        return rpc_response({"message": "Invalid period"}, 400)
    current_start, previous_start, previous_end = period_bounds(period)
    with connect() as conn:
        current_revenue = sum_amount(conn, "credit", start=current_start)
        current_expenses = sum_amount(conn, "debit", start=current_start)
        previous_revenue = sum_amount(conn, "credit", start=previous_start, end=previous_end)
        previous_expenses = sum_amount(conn, "debit", start=previous_start, end=previous_end)

    current_profit = current_revenue - current_expenses
    current_profit_percentage = (current_profit / current_revenue * 100) if current_revenue else 0
    previous_profit = previous_revenue - previous_expenses
    previous_profit_percentage = (previous_profit / previous_revenue * 100) if previous_revenue else 0

    return rpc_response({
        "revenue": {
            "value": current_revenue,
            "change": calculate_percentage_change(current_revenue, previous_revenue),
        },
        "expenses": {
            "value": current_expenses,
            "change": calculate_percentage_change(current_expenses, previous_expenses),
        },
        "profit": {
            "value": current_profit,
            "change": calculate_percentage_change(current_profit, previous_profit),
        },
        "profitPercentage": {
            "value": current_profit_percentage,
            "change": calculate_percentage_change(current_profit_percentage, previous_profit_percentage),
        },
    })


@app.post("/orpc/analyst/analytics/fleetStats")
@app.post("/api/orpc/analyst/analytics/fleetStats")
@app.post("/analyst/analytics/fleetStats")
@require_auth({"analyst"})
def orpc_analytics_fleet_stats(user):
    period = rpc_payload().get("period", "all_time")
    if period not in PERIODS:
        return rpc_response({"message": "Invalid period"}, 400)
    start = period_start(period)
    with connect() as conn:
        vehicles = rows_to_dicts(conn.execute("SELECT id, name, color FROM vehicles ORDER BY name ASC").fetchall())
        conditions = []
        params: list[Any] = []
        if start is not None:
            conditions.append("transaction_date >= ?")
            params.append(start.isoformat())
        where_clause = f"WHERE {' AND '.join(conditions)}" if conditions else ""
        credits = rows_to_dicts(conn.execute(
            f"SELECT vehicle_id, COALESCE(SUM(amount), 0) AS total FROM journal_entry_items {where_clause} AND type = 'credit' GROUP BY vehicle_id"
            if where_clause else "SELECT vehicle_id, COALESCE(SUM(amount), 0) AS total FROM journal_entry_items WHERE type = 'credit' GROUP BY vehicle_id",
            tuple(params),
        ).fetchall())
        debits = rows_to_dicts(conn.execute(
            f"SELECT vehicle_id, COALESCE(SUM(amount), 0) AS total FROM journal_entry_items {where_clause} AND type = 'debit' GROUP BY vehicle_id"
            if where_clause else "SELECT vehicle_id, COALESCE(SUM(amount), 0) AS total FROM journal_entry_items WHERE type = 'debit' GROUP BY vehicle_id",
            tuple(params),
        ).fetchall())

    credit_map = {row["vehicle_id"]: float(row["total"] or 0) for row in credits}
    debit_map = {row["vehicle_id"]: float(row["total"] or 0) for row in debits}
    vehicle_data = []
    for vehicle in vehicles:
        credit = credit_map.get(vehicle["id"], 0.0)
        debit = debit_map.get(vehicle["id"], 0.0)
        vehicle_data.append({
            "vehicleId": vehicle["id"],
            "vehicleName": vehicle["name"],
            "vehicleColor": vehicle["color"],
            "credit": credit,
            "debit": debit,
            "profit": credit - debit,
        })

    vehicle_data.sort(key=lambda row: row["profit"], reverse=True)
    return rpc_response(vehicle_data)


@app.post("/orpc/analyst/analytics/expensesStats")
@app.post("/api/orpc/analyst/analytics/expensesStats")
@app.post("/analyst/analytics/expensesStats")
@require_auth({"analyst"})
def orpc_analytics_expenses_stats(user):
    period = rpc_payload().get("period", "all_time")
    if period not in PERIODS:
        return rpc_response({"message": "Invalid period"}, 400)
    start = period_start(period)
    params: list[Any] = []
    where = ["jei.type = 'debit'"]
    if start is not None:
        where.append("jei.transaction_date >= ?")
        params.append(start.isoformat())

    with connect() as conn:
        rows = rows_to_dicts(conn.execute(
            f"""
            SELECT
                COALESCE(ec.id, 'uncategorized') AS id,
                COALESCE(ec.name, 'Uncategorized') AS name,
                COALESCE(ec.color, '6b7280') AS color,
                COALESCE(SUM(jei.amount), 0) AS amount
            FROM journal_entry_items jei
            LEFT JOIN expense_category ec ON ec.id = jei.expense_category_id
            WHERE {' AND '.join(where)}
            GROUP BY
                COALESCE(ec.id, 'uncategorized'),
                COALESCE(ec.name, 'Uncategorized'),
                COALESCE(ec.color, '6b7280')
            ORDER BY amount DESC
            """,
            tuple(params),
        ).fetchall())
    return rpc_response(rows)


@app.post("/orpc/analyst/analytics/vehicle/summaryStats")
@require_auth({"analyst"})
def orpc_vehicle_summary_stats(user):
    payload = rpc_payload()
    vehicle_id = payload.get("vehicleId")
    period = payload.get("period", "all_time")
    if not vehicle_id:
        return rpc_response({"message": "vehicleId is required"}, 400)
    if period not in VEHICLE_PERIODS:
        return rpc_response({"message": "Invalid period"}, 400)

    current_start, previous_start, previous_end = vehicle_period_bounds(period)
    with connect() as conn:
        current_revenue = sum_amount(conn, "credit", start=current_start, vehicle_id=vehicle_id)
        current_expenses = sum_amount(conn, "debit", start=current_start, vehicle_id=vehicle_id)
        previous_revenue = sum_amount(conn, "credit", start=previous_start, end=previous_end, vehicle_id=vehicle_id)
        previous_expenses = sum_amount(conn, "debit", start=previous_start, end=previous_end, vehicle_id=vehicle_id)

    current_profit = current_revenue - current_expenses
    current_profit_percentage = (current_profit / current_revenue * 100) if current_revenue else 0
    previous_profit = previous_revenue - previous_expenses
    previous_profit_percentage = (previous_profit / previous_revenue * 100) if previous_revenue else 0

    return rpc_response({
        "revenue": {
            "value": current_revenue,
            "change": calculate_percentage_change(current_revenue, previous_revenue),
        },
        "expenses": {
            "value": current_expenses,
            "change": calculate_percentage_change(current_expenses, previous_expenses),
        },
        "profit": {
            "value": current_profit,
            "change": calculate_percentage_change(current_profit, previous_profit),
        },
        "profitPercentage": {
            "value": current_profit_percentage,
            "change": calculate_percentage_change(current_profit_percentage, previous_profit_percentage),
        },
    })


@app.post("/orpc/analyst/analytics/vehicle/vehicleStats")
@require_auth({"analyst"})
def orpc_vehicle_vehicle_stats(user):
    payload = rpc_payload()
    vehicle_id = payload.get("vehicleId")
    period = payload.get("period", "all_time")
    if not vehicle_id:
        return rpc_response({"message": "vehicleId is required"}, 400)
    if period not in VEHICLE_PERIODS:
        return rpc_response({"message": "Invalid period"}, 400)

    start = vehicle_period_start(period)
    params: list[Any] = [vehicle_id]
    where = ["vehicle_id = ?"]
    if start is not None:
        where.append("transaction_date >= ?")
        params.append(start.isoformat())

    with connect() as conn:
        rows = rows_to_dicts(conn.execute(
            f"""
            SELECT strftime('%Y-%m-01T00:00:00', transaction_date) AS bucket,
                   COALESCE(SUM(CASE WHEN type = 'credit' THEN amount ELSE 0 END), 0) AS credit,
                   COALESCE(SUM(CASE WHEN type = 'debit' THEN amount ELSE 0 END), 0) AS debit
            FROM journal_entry_items
            WHERE {' AND '.join(where)}
            GROUP BY strftime('%Y-%m-01T00:00:00', transaction_date)
            ORDER BY bucket
            """,
            tuple(params),
        ).fetchall())

    stats = [
        {
            "bucket": row["bucket"],
            "credit": float(row["credit"] or 0),
            "debit": float(row["debit"] or 0),
            "profit": float(row["credit"] or 0) - float(row["debit"] or 0),
        }
        for row in rows
    ]
    return rpc_response(stats)


@app.post("/orpc/analyst/analytics/vehicle/expensesStats")
@require_auth({"analyst"})
def orpc_vehicle_expenses_stats(user):
    payload = rpc_payload()
    vehicle_id = payload.get("vehicleId")
    period = payload.get("period", "all_time")
    if not vehicle_id:
        return rpc_response({"message": "vehicleId is required"}, 400)
    if period not in VEHICLE_PERIODS:
        return rpc_response({"message": "Invalid period"}, 400)

    start = vehicle_period_start(period)
    params: list[Any] = [vehicle_id]
    where = ["jei.type = 'debit'", "jei.vehicle_id = ?"]
    if start is not None:
        where.append("jei.transaction_date >= ?")
        params.append(start.isoformat())

    with connect() as conn:
        rows = rows_to_dicts(conn.execute(
            f"""
            SELECT
                COALESCE(ec.id, 'uncategorized') AS id,
                COALESCE(ec.name, 'Uncategorized') AS name,
                COALESCE(ec.color, '6b7280') AS color,
                COALESCE(SUM(jei.amount), 0) AS amount
            FROM journal_entry_items jei
            LEFT JOIN expense_category ec ON ec.id = jei.expense_category_id
            WHERE {' AND '.join(where)}
            GROUP BY
                COALESCE(ec.id, 'uncategorized'),
                COALESCE(ec.name, 'Uncategorized'),
                COALESCE(ec.color, '6b7280')
            ORDER BY amount DESC
            """,
            tuple(params),
        ).fetchall())
    return rpc_response(rows)

